src/rep.c

```diff
@@ -0,0 +1,210 @@
+    for (R_len_t j = 1; j <= x_size; ++j, ++k) {
+      p_subscript[k] = j;
+    }
+  }
+
+  SEXP out = vec_slice_impl(x, subscript);
+
+  UNPROTECT(1);
+  return out;
+}
+
+// [[ register() ]]
+SEXP vctrs_rep(SEXP x, SEXP times) {
+  times = PROTECT(vec_cast(times, vctrs_shared_empty_int, args_times, args_empty));
+
+  if (vec_size(times) != 1) {
+    stop_rep_times_size();
+  }
+
+  const R_len_t times_ = r_int_get(times, 0);
+
+  SEXP out = vec_rep(x, times_);
+
+  UNPROTECT(1);
+  return out;
+}
+
+// -----------------------------------------------------------------------------
+
+static SEXP vec_rep_each_uniform(SEXP x, R_len_t times) {
+  if (times < 0) {
+    if (times == NA_INTEGER) {
+      stop_rep_each_times_missing(1);
+    } else {
+      stop_rep_each_times_negative(1);
+    }
+  }
+
+  const R_len_t x_size = vec_size(x);
+
+  const long long temp_size = (long long) x_size * times;
+  if (temp_size > INT_MAX) {
+    stop_rep_size_oob(temp_size);
+  }
+
+  const R_len_t size = (R_len_t) temp_size;
+
+  SEXP subscript = PROTECT(Rf_allocVector(INTSXP, size));
+  int* p_subscript = INTEGER(subscript);
+
+  R_len_t k = 0;
+
+  for (R_len_t i = 1; i <= x_size; ++i) {
+    for (R_len_t j = 0; j < times; ++j, ++k) {
+      p_subscript[k] = i;
+    }
+  }
+
+  SEXP out = vec_slice_impl(x, subscript);
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP vec_rep_each_impl(SEXP x, SEXP times, const R_len_t times_size) {
+  const R_len_t x_size = vec_size(x);
+
+  if (x_size != times_size) {
+    stop_recycle_incompatible_size(times_size, x_size, args_times);
+  }
+
+  const int* p_times = INTEGER_RO(times);
+
+  long long temp_size = 0;
+  for (R_len_t i = 0; i < times_size; ++i) {
+    const int elt_times = p_times[i];
+
+    if (elt_times < 0) {
+      if (elt_times == NA_INTEGER) {
+        stop_rep_each_times_missing(i + 1);
+      } else {
+        stop_rep_each_times_negative(i + 1);
+      }
+    }
+
+    temp_size += elt_times;
+  }
+
+  if (temp_size > INT_MAX) {
+    stop_rep_size_oob(temp_size);
+  }
+
+  const R_len_t size = (R_len_t) temp_size;
+
+  SEXP subscript = PROTECT(Rf_allocVector(INTSXP, size));
+  int* p_subscript = INTEGER(subscript);
+
+  R_len_t k = 0;
+
+  for (R_len_t i = 1; i <= x_size; ++i) {
+    const int elt_times = p_times[i - 1];
+
+    for (R_len_t j = 0; j < elt_times; ++j, ++k) {
+      p_subscript[k] = i;
+    }
+  }
+
+  SEXP out = vec_slice_impl(x, subscript);
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP vec_rep_each(SEXP x, SEXP times) {
+  times = PROTECT(vec_cast(times, vctrs_shared_empty_int, args_times, args_empty));
+
+  const R_len_t times_size = vec_size(times);
+
+  SEXP out;
+
+  if (times_size == 1) {
+    const R_len_t times_ = r_int_get(times, 0);
+    out = vec_rep_each_uniform(x, times_);
+  } else {
+    out = vec_rep_each_impl(x, times, times_size);
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+// [[ register() ]]
+SEXP vctrs_rep_each(SEXP x, SEXP times) {
```

Comment:
Should we start the convention that higher-level functions are always on top? It helps to understand the context in which lower-level functions are used before diving into the implementation. This makes it easier to understand a file.

For example in the rep-each section we'd start with `vctrs_rep_each()`, then `vec_rep_each()`, then `vec_rep_each_uniform()`, then the internal helpers for rep-each-uniform, then `vec_rep_each_impl()`, etc.